### PR TITLE
Spring boot criteria- query usage. Added ability to search books by parameters

### PR DESCRIPTION
**Pull request by YaroslavRadevychVynnytskyi**

### CODE REVIEW

File: src/main/java/application/bookstore/repository/book/BookSpecificationProviderManager.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+    public BookSpecificationProviderManager(List<SpecificationProvider<Book>>
+                                                    bookSpecificationProviders) {
+        this.bookSpecificationProviders = bookSpecificationProviders;
+    }
+
+    @Override
+    public SpecificationProvider<Book> getSpecificationProvider(String key) {
+        return bookSpecificationProviders.stream()
+                .filter(p -> p.getKey().equals(key))
```
Comment (okuzan) [inline src/main/java/application/bookstore/repository/book/BookSpecificationProviderManager.java]: avoid short one letter naming even in stream chains, readability is more important that a second you save not writing out that full name

File: src/main/java/application/bookstore/repository/book/spec/TitleSpecificationProvider.java [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+public class TitleSpecificationProvider implements SpecificationProvider<Book> {
+    @Override
+    public String getKey() {
+        return "title";
+    }
+
+    @Override
+    public Specification<Book> getSpecification(String[] params) {
+        return (root, query, criteriaBuilder) ->
+                root.get("title").in(Arrays.stream(params).toArray());
```
Comment (okuzan) [inline src/main/java/application/bookstore/repository/book/spec/TitleSpecificationProvider.java]: let's extract this literal to a constant

**Review by okuzan — CHANGES_REQUESTED** [review]

File: src/main/java/application/bookstore/repository/book/BookSpecificationBuilder.java [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    public Specification<Book> build(BookSearchParameters searchParameters) {
+        Specification<Book> specification = Specification.where(null);
+        if (searchParameters.titles() != null && searchParameters.titles().length > 0) {
+            specification = specification.and(specificationProviderManager
+                    .getSpecificationProvider("title")
+                    .getSpecification(searchParameters.titles()));
+        }
+        if (searchParameters.authors() != null && searchParameters.authors().length > 0) {
+            specification = specification.and(specificationProviderManager
+                    .getSpecificationProvider("author")
```
Comment (MentorMA) [inline src/main/java/application/bookstore/repository/book/BookSpecificationBuilder.java]: For strings you can use constants and I see you already have constant author and title in other classes

**Review by MentorMA — CHANGES_REQUESTED** [review]

File: src/main/java/application/bookstore/repository/book/BookSpecificationBuilder.java [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+import application.bookstore.repository.SpecificationProviderManager;
+import lombok.RequiredArgsConstructor;
+import org.springframework.data.jpa.domain.Specification;
+import org.springframework.stereotype.Component;
+
+@RequiredArgsConstructor
+@Component
+public class BookSpecificationBuilder implements SpecificationBuilder<Book> {
+    private static final String TITLE = "title";
+    private static final String AUTHOR = "author";
```
Comment (MentorMA) [inline src/main/java/application/bookstore/repository/book/BookSpecificationBuilder.java]: You already have these constants in SpecificationProvider classes, don`t duplicate code

**Review by MentorMA — CHANGES_REQUESTED** [review]

**Pull request merged**